### PR TITLE
chore(utxo-lib): wrap verifySignature checks for zcash as unsafe

### PR DESCRIPTION
**Pull request by davidkaplanbitgo**
For Zcash transactions, we do not have to have non-witness UTXO data for non-segwit transactions because zcash hashes the value directly. Thus, it is unnecessary to have the previous transaction hash on the unspent. We did this for signing inputs and we need to do it for validating inputs as well.

BG-66521

### CODE REVIEW

File: modules/utxo-lib/src/bitgo/zcash/ZcashPsbt.ts [hunk 1]
```diff
@@ -128,14 +129,30 @@ export class ZcashPsbt extends UtxoPsbt<ZcashTransaction<bigint>> {
   // transactions because zcash hashes the value directly. Thus, it is unnecessary to have
   // the previous transaction hash on the unspent.
   signInput(inputIndex: number, keyPair: Signer, sighashTypes?: number[]): this {
-    (this as any).__CACHE.__UNSAFE_SIGN_NONSEGWIT = true;
+    this.setUnsafeSignNonSegwit(true);
```
Comment (brandonblack) [inline modules/utxo-lib/src/bitgo/zcash/ZcashPsbt.ts:132]: Let's do this as a `with` now that we're making it cleaner.

**Review by brandonblack — CHANGES_REQUESTED** [review]

File: modules/utxo-lib/src/bitgo/zcash/ZcashPsbt.ts (hunk 1 above)
Comment (davidkaplanbitgo) [reply modules/utxo-lib/src/bitgo/zcash/ZcashPsbt.ts:132]: Is it depreciated? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/with

**Review by lcovar — DISMISSED** [review]

File: modules/utxo-lib/src/bitgo/zcash/ZcashPsbt.ts [hunk 2]
```diff
@@ -128,9 +129,17 @@ export class ZcashPsbt extends UtxoPsbt<ZcashTransaction<bigint>> {
...
+
+  validateSignaturesOfInput(inputIndex: number, validator: ValidateSigFunction, pubkey?: Buffer): boolean {
+    return this.withUnsafeSignNonSegwitTrue(super.validateSignaturesOfInput.bind(this, inputIndex, validator, pubkey));
+  }
+
+  private withUnsafeSignNonSegwitTrue(callback: () => any): any {
     (this as any).__CACHE.__UNSAFE_SIGN_NONSEGWIT = true;
     try {
-      return super.signInput(inputIndex, keyPair, sighashTypes);
+      return callback();
```
Comment (brandonblack) [inline modules/utxo-lib/src/bitgo/zcash/ZcashPsbt.ts:142]: hate to be that guy, but callback is a confusing name here. `fn`, `func`, `call`, something that doesn't already have an accepted meaning different from what we're doing.

**Review by brandonblack — CHANGES_REQUESTED** [review]

File: modules/utxo-lib/src/bitgo/zcash/ZcashPsbt.ts [hunk 3]
```diff
@@ -128,9 +129,17 @@ export class ZcashPsbt extends UtxoPsbt<ZcashTransaction<bigint>> {
...
   // the previous transaction hash on the unspent.
   signInput(inputIndex: number, keyPair: Signer, sighashTypes?: number[]): this {
+    return this.withUnsafeSignNonSegwitTrue(super.signInput.bind(this, inputIndex, keyPair, sighashTypes));
+  }
+
+  validateSignaturesOfInput(inputIndex: number, validator: ValidateSigFunction, pubkey?: Buffer): boolean {
+    return this.withUnsafeSignNonSegwitTrue(super.validateSignaturesOfInput.bind(this, inputIndex, validator, pubkey));
+  }
+
+  private withUnsafeSignNonSegwitTrue(fn: () => any): any {
```
Comment (brandonblack) [inline modules/utxo-lib/src/bitgo/zcash/ZcashPsbt.ts:139]:
```suggestion
  private withUnsafeSignNonSegwitTrue<T>(fn: () => T): T {
```
Does this work to improve type checking?

**Review by brandonblack — CHANGES_REQUESTED** [review]

File: modules/utxo-lib/src/bitgo/zcash/ZcashPsbt.ts (hunk 3 above)
Comment (davidkaplanbitgo) [reply modules/utxo-lib/src/bitgo/zcash/ZcashPsbt.ts]: Yes - this works - and a lot better than how it was before

**Review by brandonblack — APPROVED** [review]
Wow, that turned out clean-a-f :-D

**Pull request merged**